amlb/framework_definitions.py

```diff
@@ -0,0 +1,182 @@
+    return frameworks
+
+
+def load_and_merge_framework_definitions(frameworks_file: Union[str, List[str]]) -> Namespace:
+    """ Load and merge the framework file(s), does not allow duplicate definitions. """
+    log.info("Loading frameworks definitions from %s.", frameworks_file)
+    if not isinstance(frameworks_file, list):
+        frameworks_file = [frameworks_file]
+
+    definitions_by_file = [config_load(file) for file in frameworks_file]
+    for d1, d2 in itertools.combinations([set(dir(d)) for d in definitions_by_file], 2):
+        if d1.intersection(d2) != set():
+            raise ValueError(f"Duplicate entry '{d1.intersection(d2).pop()}' found.")
+    return Namespace.merge(*definitions_by_file)
+
+
+def sanitize_definitions(frameworks: Namespace):
+    """ Normalize names, add name field, remove invalid extensions. """
+    add_and_normalize_names(frameworks)
+    remove_frameworks_with_unknown_parent(frameworks)
+    remove_self_reference_extensions(frameworks)
+
+
+def sanitize_and_add_defaults(frameworks, resource):
+    sanitize_definitions(frameworks)
+
+    # `module` is the only field that should have a default
+    # based on the parent. For that reason we add it before
+    # we update children with their parent fields.
+    for _, framework in frameworks:
+        if "extends" not in framework:
+            add_default_module(framework, resource.config)
+    update_frameworks_with_parent_definitions(frameworks)
+
+    add_defaults_to_frameworks(frameworks, resource)
+
+
+def add_and_normalize_names(frameworks: Namespace):
+    """ Converts each framework definition to lowercase and adds a 'name' field. """
+    framework_names = dir(frameworks)
+    for name in framework_names:
+        framework = frameworks[name]
+        framework.name = name.lower()
+        if name.lower() != name:
+            del frameworks[name]
+            frameworks[name.lower()] = framework
+        if "extends" in framework:
+            framework.extends = framework.extends.lower()
+
+
+def add_default_module(framework, config):
+    if "module" not in framework:
+        framework.module = f"{config.frameworks.root_module}.{framework.name}"
+
+
+def add_default_version(framework):
+    if "version" not in framework:
+        framework.version = "latest"
+
+
+def add_default_setup_args(framework):
+    if "setup_args" in framework:
+        framework.setup_args = [framework.setup_args]
+    else:
+        framework.setup_args = [framework.version]
+        if "repo" in framework:
+            framework.setup_args.append(framework.repo)
+
+
+def add_default_setup_script(framework, resource):
+    if "setup_script" not in framework:
+        framework.setup_script = None
+    else:
+        framework.setup_script = framework.setup_script.format(
+            module=framework.module,
+            **resource._common_dirs,
+        )
+
+
+def add_default_setup_cmd(framework, resource):
```

Comment:
```
    """ Defines default setup_cmd and _setup_cmd, interpolate commands if necessary.

    The default values are `None`.
    In case a setup_cmd is defined, the original definition is saved to `_setup_cmd`.
    The new `setup_cmd` will be a list of commands, where each command has the
    directories, package manager and python binary interpolated.
    `_setup_cmd` will be used for setup inside containers.
    `setup_cmd` will be used when running locally (or on an Amazon image).
    """
Does that sound right?
```

Reply:
Added suggestion below. I think you can edit and/or commit directly.